src/docs/arch-rts.dox

```diff
@@ -110,4 +109,154 @@ Update list: U3 (30)\n
+generation will be initialized after RTS and the updates in the cache would be problematic for other
+parts of the code if they had a "legitimate" transaction ID.
+
+For example, suppose we restore an update from the history store with a transaction ID of 2000 and
+add it back to the update list. Later, after recovery, a new transaction starts with a
+minimum/maximum snapshot of (5, 10). This would not see the update because it's transaction ID
+(2000) is greater than the maximum snapshot of the new transaction. By removing the transaction ID,
+the update becomes globally visible, solving this problem.
+
+We treat all data as visible when RTS is called from the API. Since we clear transaction IDs during
+RTS, we could also do this during the rollback that happens during startup and shutdown.
+
+@section operations-on-the-history-store Operations on the history store
+
+RTS touches the history store for a lot of things, primarily for deleting, but it might also need to
+read an update to move it to the data store.
+
+The deletion code is called from the general update aborting code, where we might need to delete the
+first stable update we found when it came from the history store. There's also a somewhat subtle
+case where a new key can be added to a page, and then the page checkpointed, leading to a situation
+where we have updates to the key in the history store but not the disk image.
+
+History store deletions are done using a standard history store cursor, starting at the highest
+possible start timestamp, and removing keys from newest to oldest until we hit an update with a
+start timestamp larger than the rollback timestamp.
+
+Note: in-memory databases don't have a history store, so none of this section will apply.
+
+@section iterating-over-b-trees Iterating over B-Trees
+
+This is one of the simplest parts of the RTS subsystem, as a metadata cursor is specifically
+designed for these sorts of "iterate over a B-Tree" operations. We use this functionality for RTS,
+with the caveat that we tweak a few options so that corrupt or missing files don't raise errors. The
+reasoning here is that RTS is orthogonal to salvage, and might be called as part of salvage, so we
+simply need to "do our best". We do log the issues we find, however.
+
+Less obviously, the history store needs one final inspection regardless of whether any other B-Trees
+were touched. This is because we usually rely on getting to the history store "via" the data store,
+but if we don't touch a B-Tree, that doesn't mean there aren't updates in the history store after
+the stable timestamp.
+
+@section iterating-over-pages Iterating over pages
+
+RTS tries to avoid looking at a page if it can. There are two ways we can avoid instantiating a
+page: if it was truncated at/before the stable timestamp, and if the page state is anything other
+than on-disk. If neither of these are satisfied, we need to instantiate the page to look at it, but
```

Comment:
If the page state is other than on-disk means it is already in the cache, so there is no instantiation here. Are you explaining the page with the on-disk state that does not have unstable updates will not be instantiated?

Reply:
It's the latter, and I think I wrote this the wrong way around for some reason - I believe it should say "... or if the page state is on-disk. If ..."

Reply:
Draft fix in 5d27bf2, let me know if it's OK.